### DISCUSSION TITLE
ST7567S 12864 (128x64) I2C COG LCD

### DISCUSSION THREAD

**Discussion by JamoDevNich** [olikraus/u8g2#2560] [General]
Had a quick browse through some of the existing issues where this LCD was mentioned, and tried a few of the constructors to see what works.

![image](https://github.com/user-attachments/assets/b48d90c5-0d39-482f-8170-805af9a4963a)

All of these have been tested on an ESP32 (not a newer S2, S3, or C3) @ 3.3v using hardware SPI

1) 
    ```c++
    U8G2_ST7567_OS12864_F_HW_I2C u8g2(U8G2_R2, U8X8_PIN_NONE, 22, 21);
    ```

    - Works after specifying `u8g2.setI2CAddress(0x3F * 2)`; prior to `u8g2.begin();`
    - Contrast levels seem excessive
    - About 20px of space missing on the right??

&nbsp;

2)
    ```c++
    U8G2_ST7567_HEM6432_F_HW_I2C u8g2(U8G2_R2, U8X8_PIN_NONE, 22, 21);
    ```

    - Works in a small 64 x 32 region of the display, not usable

&nbsp;

3)
    ```c++
    U8G2_ST7567_ENH_DG128064I_F_HW_I2C u8g2(U8G2_R2, 22, 21);
    ```

    - Works after specifying `u8g2.setI2CAddress(0x3F * 2)`; prior to `u8g2.begin();`
    - Contrast is OK
    - About 20px of space missing on the right??

&nbsp;

4)
    ```c++
    U8G2_ST7567_ENH_DG128064_F_HW_I2C u8g2(U8G2_R2, 22, 21);
    ```

    - Works after specifying `u8g2.setI2CAddress(0x3F * 2)`; prior to `u8g2.begin();`
    - 5px space missing on the left
    - Contrast is OK
    - About 15px of space missing on the right??

&nbsp;

5)
    ```c++
    U8G2_ST7567_JLX12864_F_HW_I2C u8g2(U8G2_R2, U8X8_PIN_NONE, 22, 21);
    ```

    - Works after specifying `u8g2.setI2CAddress(0x3F * 2)`; prior to `u8g2.begin();`
    - Display occasionally shows glitching
    - Contrast set after begin() with `u8g2.setContrast(160);`

&nbsp;

Hope this is helpful to someone.

**Comment by olikraus** [comment]
Thanks for your report. The mapping between the display and the controller memory differs from display to display, thats why there might be missing pixel or space between pixel and border. Also the contrast is always specific to the display itself.

**Reply by flybabo** [reply]
Do you believe that the display unit has a fundamental quality problem?

**Reply by JamoDevNich** [reply]
There are no quality issues with mine personally, as the main dev said the mapping of the controller and the display are different, the constructors take that into account.

I use this display with the constructor `U8G2_ST7567_ENH_DG128064I_F_HW_I2C` and the setup to correct the small discrepancies caused by using the wrong constructor:

```
void setup(void) {
  u8g2.setI2CAddress(0x3F * 2);
  u8g2.begin();
  u8g2.setContrast(230);
}
```

I have not been able to reproduce the missing 20px on the right since those initial tests - it's worked just fine since

**Comment by Pepitof** [comment]
Thank you very much for your input.
I use the exact same LCD module (GM12864-59N) and it works perfectly with that constructor (U8G2_ST7567_JLX12864_F_HW_I2C), but
I haven't had the glitching problem you mentioned.
In case it helps, I added 3k pull-up resistors on SDA and SCL, and removed zener diodes D1 and D2 from the LCD module.
